Introduce platform and board hooks

nashif opened the pull request:
Introduce hook infrastructure for platform/board initialization. Replaces existing hooks implemented for few architectures and unifies them and adds hooks to replace common platform code currently implemented using SYS_INIT


resolves https://github.com/zephyrproject-rtos/zephyr/issues/75907

nashif committed:
init: introduce soc and board hooks

Introduce soc and board hooks to replace arch specific code
and replace usages of SYS_INIT for platform initialization.

include/zephyr/platform/hooks.h introduces the hooks to be implemented
by boards and SoCs.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

gmarull commented on kernel/init.c:
```diff
@@ -550,6 +555,9 @@ FUNC_NORETURN void z_cstart(void)
 
+#if CONFIG_PLATFORM_EARLY_INIT
+	platform_early_initialize();
+#endif
```
wondering if this is early enough (considering EARLY exists and it is called just before entering the C domain). For example, z_device_state_init should be called after. Also, is `arch_kernel_init()` initializing something for the Kernel, or is it more like arch initial code (also required by early hooks)?

nashif replied:
right now I am just trying to make this work exactly the same way as before, architecture initialization might have other caveats with multi core and SMP and other memory related activities. We probably can eliminate EARLY and make this an arch hook as well. Anyways, the positioning is subject to optimizations and change, this is more about the concept...

gmarull replied:
I think the concept is ok, but we should likely merge something that doesn't change in the near term, even if it implies some slight behavior changes.

dleach02 replied:
Why couldn't the invocation of a platform_early_initialize() just be a function run at PRE_KERNEL_1 priority 1 (or zero)?

tbursztyka replied:
If you can hard-code the function call, you'll gain one init_entry less in ROM. (ok that's a very little amount)

SYS_INIT it meant to disappear, along with its hard-coded priority anyway.

The most important part of this PR is not where these functions are called: it could be easily fixed or rendered more flexible (whit what's going to replace SYS_INIT) since in kernel/init.c.  The API is what matters

gmarull commented on include/zephyr/platform/platform.h:
```diff
@@ -0,0 +1,13 @@
+#define ZEPHYR_INCLUDE_PLATFORM_PLATFORM_H_
+
+
+void platform_early_initialize(void);
+void platform_late_initialize(void);
```
since these are called internally, they should not be exposed in a public header. we should likely think about internal headers... now everything is under `include/zephyr`, sometimes you can see `header_internal.h`, but it's not a strict rule.

nashif replied:
> since these are called internally, they should not be exposed in a public header. we should likely think about internal headers... now everything is under `include/zephyr`, sometimes you can see `header_internal.h`, but it's not a strict rule.

internal to what? The call is internal, but the implementation can be external to zephyr, so we need to let implementers know what API to implement. So we need to somehow expose this as an interface (public) that can be implemented and not called somehow. We will have the same issue with other hooks, so we need to figure out how to position it and expose it.

gmarull replied:
Only Kernel is meant to call these hooks, so placing them in `zephyr/platform.h` will make people feel they can include that file and call the functions somewhere else.

nashif replied:
yes, I understand, let me think about it a bit more.

nashif replied:
some googling reveals others with similar approach :)

https://github.com/apache/nuttx/blob/master/include/nuttx/board.h

https://github.com/apache/nuttx/blob/a73f5f86dcb05ec98d66f70966f02624f2e2c9f3/include/nuttx/board.h#L103:

> * Public Function Prototypes
>  *
> * These are all standard board interfaces that are exported from board-
>  * specific logic to OS internal logic.  These should never be accessed
>  * directly from application code but may be freely used within the OS
>  *

We could have something similar,

gmarull replied:
Zephyr looking at NuttX :open_mouth: (just joking, I think NuttX has many well-developed concepts)

tbursztyka commented:
Missed that PR. This is a good idea, this would normalize SOC initialization across zephyr. 

Not sure we want to keep the level names in the functions however. Most of SOCs uses soc[_.]1_init() and optionally soc[_.]1_config() 
Thus what about platform_init()/platform_config() ? (or, actually, shouldn't we keep soc_init()/soc_config() instead of platform as it targets code located in soc/ ?)

tbursztyka commented:
Quickly checking SYS_INIT() usage in soc/, there are 145 ones. Most of them on PRE_KERNEL_1 whether it is _init or _config 
To be more precise:
- 6 at EARLY
- 119 at PRE_KERNEL_1
- 11 at PRE_KERNEL_2
- 8 at POST_KERNEL
- 1 at APPLICATION

There are a few use case that do not land into that scheme like power initialization sometimes occurring at PRE_KERNEL_2, or some secondary SOC or "slave" (whatever this is). But these are the least present ones, they could keep using SYS_INIT().

tbursztyka commented:
@nashif will you continue on this?

nashif commented:
> @nashif will you continue on this?

yes

nashif committed:
arch: arm: replace PLATFORM_SPECIFIC_INIT with PLATFORM_RESET_HOOK

Use generic hook infrastrucutre instead of custom Kconfig and hooks for
ARM.

Replace z_arm_platform_init() with platform_reset().

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: arc: replace ARC_EARLY_SOC_INIT with PLATFORM_RESET_HOOK

Use generic hook infrastrucutre instead of custom Kconfig and hooks for
ARC.

Replace soc_early_asm_init_percpu() with platform_reset()

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: arm: select PLATFORM_RESET_HOOK if is PLATFORM_SPECIFIC_INIT set

Temporary until usage of PLATFORM_SPECIFIC_INIT is removed in modules.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

fabiobaltieri commented on arch/arc/core/prep_c.c:
```diff
@@ -121,6 +122,10 @@ extern FUNC_NORETURN void z_cstart(void);
 
 void z_prep_c(void)
 {
+#if defined(CONFIG_PLATFORM_PREP_HOOK)
```
would it make sense to have weak functions instead of config options? it's already the case for other arch functions (`arch_system_halt`, `arch_spin_relax`...) would save few options, it's not like you may ever want to implement it and turn the option off anyway

henrikbrixandersen replied:
I think having weak defaults makes sense.

nashif replied:
the `__weak` approach is mentioned in the RFC as an option. We had many issues with weak functions in the past (defined at multiple levels (arch, soc, board) so you never knew which one you got etc, maybe it is not the case here. I also remember some compilers/architectures having problems with weak symbols the reason why existing hooks did not use them. Anyways, lets see if there is more feedback input to that and we can switch to weak definitions.

tbursztyka replied:
this way adds a lot of #ifdef, __weak would be nice yes (if there aren't any issues with it)

nashif replied:
at least for one of the hooks and for some platforms, the hooks is being disabled in some samples...

```
samples/drivers/ipm/ipm_mcux/remote/prj.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp/remote/prj.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp_rsc_table/boards/colibri_imx7d_mcimx7d_m4.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp_rsc_table/boards/stm32mp157c_dk2.conf:CONFIG_PLATFORM_RESET_HOOK=n
```

need to find a out why and figure a better way to deal with this using weak symbols...

tbursztyka reviewed (commented):
Pehaps we should take some time to get good hook names.
Should we really use short names? Why not full names (i.e. s/prep/preparation)
And trying to be more specific also: platform_boot_preparation() ?
For instance, I prefer platform_post_kernel_init() rather than platform_late_init() which is less self-documenting. It's all subjective of course, and most probably there are better proposals.

nashif commented:
> Pehaps we should take some time to get good hook names.
> Should we really use short names? Why not full names (i.e. s/prep/preparation)
> And trying to be more specific also: platform_boot_preparation() ?
> For instance, I prefer platform_post_kernel_init() rather than platform_late_init() which is less self-documenting. It's all subjective of course, and most probably there are better proposals.

open for suggestions. we probably do not want to make this kernel centric, as more things happen beside kernel initialization  between the hooks.

ithinuel commented:
At first I was a bit worried by the names but the documentation in hook.h clearly states that these should not be called from the used application so I’m fine with them as is 👍.

carlescufi commented:
Architecture WG:

- Proposal is welcome in general
- @henrikbrixandersen suggests changing `platform_` to `soc_` since we don't really have a clear concept of "platform" in Zephyr
- @ithinuel points to a PR with similar requirements: https://github.com/zephyrproject-rtos/zephyr/pull/71667 which uses a similar approach
- Kconfig vs weak functions: @nashif outlines pros and cons
- General consensus seems to form in favor of Kconfig instead of weak functions

nashif commented:
> * @henrikbrixandersen suggests changing `platform_` to `soc_` since we don't really have a clear concept of "platform" in Zephyr

addressed

henrikbrixandersen commented on include/zephyr/platform/hooks.h:
```diff
@@ -0,0 +1,74 @@
+ * @brief Platform hooks
+ *
+ * This header file contains function prototypes for the interfaces between
+ * zephyr architecture and initialization code and the platform pecific logic
```
```suggestion
 * zephyr architecture and initialization code and the platform specific logic
```

henrikbrixandersen commented on include/zephyr/platform/hooks.h:
```diff
@@ -0,0 +1,74 @@
+
+
+/**
+ * @brief Platform hook executed at the beginning of the reset vector.
```
```suggestion
 * @brief SoC hook executed at the beginning of the reset vector.
```

Same change should be done for the remaining SoC hooks.

henrikbrixandersen commented on include/zephyr/platform/hooks.h:
```diff
@@ -0,0 +1,74 @@
+/**
+ * @brief Platform hook executed at the beginning of the reset vector.
+ *
+ * This hook is implemented by the SoC and can be used to perform any
+ * SoC-specific
```
Any SoC-specific... what? :) Same comment goes for the remaining hook definitions.

henrikbrixandersen commented on include/zephyr/platform/hooks.h:
```diff
@@ -0,0 +1,74 @@
+/*
+ * @brief Additional platform hook executed after the kernel starts.
+
+ * This is called after the kernel has started, but before the main. This hook
```
```suggestion
 * This is called after the kernel has started, but before the main function is called. This hook
```

henrikbrixandersen commented on kernel/Kconfig.init:
```diff
@@ -0,0 +1,45 @@
+	  Run an early SoC reset hook.
+
+	  A custom hook SoC_reset() is executed at the beginning of the
+	  startup code (__start). SoC_reset() must be implemented by the SoC.
```
`SoC_reset()`?

henrikbrixandersen commented on kernel/Kconfig.init:
```diff
@@ -0,0 +1,45 @@
+	  Run an early SoC preparation hook.
+
+	  A custom hook SoC_prep() is executed at the beginning of the
+	  c prep code (prep_c). SoC_prep() must be implemented by the SoC.
```
`SoC_prep()`?

henrikbrixandersen commented on soc/nxp/imxrt/imxrt6xx/Kconfig:
```diff
@@ -6,7 +6,8 @@ config SOC_MIMXRT685S_CM33
 	select CPU_CORTEX_M_HAS_DWT
 	select CLOCK_CONTROL
-	select PLATFORM_SPECIFIC_INIT
+	select CODE_DATA_RELOCATION_SRAM if FLASH_MCUX_FLEXSPI_XIP
```
Was this change intended?

nashif replied:
no, rebase mess. removed.

henrikbrixandersen reviewed (changes requested):
A few comments regarding the provided documentation strings, but otherwise this looks like a good foundation.

pillo79 commented on soc/nxp/kinetis/ke1xz/Kconfig:
```diff
@@ -11,5 +11,4 @@ config SOC_SERIES_KE1XZ
 	select CLOCK_CONTROL
 	select HAS_MCUX
-	select PLATFORM_SPECIFIC_INIT
-	select HAS_PM
```
I think this is a cut and paste error - why is HAS_PM being removed for this soc only?

nashif replied:
yeah, rebase issue, fixed.

nandojve reviewed (approved):
* Atmel
* GD32

decsny reviewed (approved):
very good, thanks for handling the NXP socs conversion

decsny commented:
is there an estimated timeframe we should target to have the boards converted to this scheme by once this merges?

nashif commented:
> is there an estimated timeframe we should target to have the boards converted to this scheme by once this merges?

plan is to do mass migration using a few scripts @tbursztyka  created. Some boards that need special attention will probably remain and need to be done individually.

kartben commented on kernel/Kconfig.init:
```diff
@@ -0,0 +1,45 @@
+	  A custom hook soc_prep_hook() is executed at the beginning of the
+	  c prep code (prep_c). soc_prep_hook() must be implemented by the SoC.
+
+config SOC_EARLY_INIT_HOOK
```
Can be a follow-up commit but would be nice if all the helps would document the name of the hook function like you did in the first two above?

bjarki-andreasen reviewed (approved):
Nice work!

ajf58 commented on arch/arm/core/Kconfig:
```diff
@@ -166,11 +166,15 @@ config RUNTIME_NMI
+	default y if SOC_RESET_HOOK
+	select DEPRECATED
 	help
 	  The platform specific initialization code (z_arm_platform_init) is
 	  executed at the beginning of the startup code (__start).
```
This comment is now very misleading. `z_arm_platform_init` isn't ever called anymore.  I think this is a bug.

ajf58 commented:
https://github.com/zephyrproject-rtos/zephyr/issues/78386
